File: matlab/metasignal_mat/tutorials/02_computing_measures.m
%% Tutorial 2 — Computing All 20 Measures
% Detailed walkthrough of each block of the 20-measure array, plus how to
% call individual measure functions directly.
%
% Measure order:
%   [1-5]   Absolute sensitivity: meta-d', AUC2, Gamma, Phi, DeltaConf
%   [6-10]  Efficiency ratios:   M-Ratio, AUC2-Ratio, Gamma-Ratio, Phi-Ratio, DeltaConf-Ratio
%   [11-15] Efficiency diffs:    M-Diff, AUC2-Diff, Gamma-Diff, Phi-Diff, DeltaConf-Diff
%   [16-17] Noise models:        meta-noise, meta-uncertainty
%   [18-20] Type-1 / raw:        d', Criterion, mean Confidence

%% Setup
clear; close all; clc

root_dir = fileparts(fileparts(mfilename('fullpath')));
addpath(genpath(fullfile(root_dir, 'helperFunctions')));

variable_names = {'meta-d''', 'AUC2', 'Gamma', 'Phi', 'DeltaConf', ...
    'M-Ratio', 'AUC2-Ratio', 'Gamma-Ratio', 'Phi-Ratio', 'DeltaConf-Ratio', ...
    'M-Diff', 'AUC2-Diff', 'Gamma-Diff', 'Phi-Diff', 'DeltaConf-Diff', ...
    'meta-noise', 'meta-uncertainty', 'd''', 'Criterion', 'Confidence'};

rng(0);
n_trials = 400;
nRatings = 4;
accuracy = 0.78;

stim = randi([0,1], n_trials, 1);
resp = stim;
flip = rand(n_trials,1) > accuracy;
resp(flip) = 1 - resp(flip);
correct = (stim == resp);
conf = zeros(n_trials,1);
conf( correct) = randi([3, nRatings], sum( correct), 1);
conf(~correct) = randi([1,         2], sum(~correct), 1);

fprintf('Data ready: %d trials, nRatings=%d\n', n_trials, nRatings);

%% Block 1 — Absolute sensitivity (indices 1–5)
% These five measures ask: how well does confidence track accuracy?
% None of them are normalised by what an ideal observer would achieve.

meas = compute_all_measures(stim, resp, conf, nRatings);

labels = {'meta-d''', 'AUC2', 'Gamma', 'Phi', 'DeltaConf'};
fprintf('\nBlock 1 — Absolute sensitivity:\n');
for i = 1:5
    fprintf('  %-12s = %.4f\n', labels{i}, meas(i));
end

%% Block 2 & 3 — Efficiency ratios and differences (indices 6–15)
% Normalise observed metacognition by the expected performance of an ideal
% observer with the same d'. Removes spurious dependence on task difficulty.
% SDTexpectConf computes the expected nR_S1/nR_S2 under ideal SDT.

% AUC2 observed vs expected
[auc2_obs, auc2_ratio, auc2_diff] = SDTtype2AUC(stim, resp, conf, nRatings);

fprintf('\nBlock 2 & 3 — Efficiency measures (AUC2 example):\n');
fprintf('  AUC2 observed = %.4f\n', auc2_obs);
fprintf('  AUC2-Ratio    = %.4f  (obs / ideal)\n', auc2_ratio);
fprintf('  AUC2-Diff     = %.4f  (obs - ideal)\n', auc2_diff);

%% Individual measure functions
% You can also call each measure directly without going through
% compute_all_measures.

[gamma, gamma_ratio, gamma_diff]           = SDTgamma(stim, resp, conf, nRatings);
[phi,   phi_ratio,   phi_diff]             = SDTphi(stim, resp, conf, nRatings);
[dc,    dc_ratio,    dc_diff]              = SDTdeltaConf(stim, resp, conf, nRatings);

fprintf('\nIndividual functions:\n');
fprintf('  Gamma      = %.4f\n', gamma);
fprintf('  Phi        = %.4f\n', phi);
fprintf('  DeltaConf  = %.4f\n', dc);

% meta-d' via MLE (type2_SDT_MLE)
output  = type2_SDT_MLE(stim, resp, conf, nRatings, [], 1);
fprintf('  meta_da    = %.4f\n', output.meta_da);
fprintf('  M_ratio    = %.4f\n', output.M_ratio);

%% Block 4 — Meta-noise and meta-uncertainty (indices 16–17)
[metaNoise, ~]     = compute_metaNoise(stim, resp, conf, nRatings);
metaUncertainty    = compute_metaUncertainty(stim, resp, conf, nRatings);

fprintf('\nBlock 4 — Noise-based measures:\n');
fprintf('  meta-noise       = %.4f\n', metaNoise);
fprintf('  meta-uncertainty = %.4f\n', metaUncertainty);

%% Full 20-measure summary
fprintf('\n%s\n', repmat('-', 1, 42));
fprintf('%-6s  %-22s  %s\n', 'Index', 'Measure', 'Value');
fprintf('%s\n', repmat('-', 1, 42));
for i = 1:20
    if isnan(meas(i))
        fprintf('[%2d]   %-22s  NaN\n', i, variable_names{i});
    else
        fprintf('[%2d]   %-22s  %10.4f\n', i, variable_names{i}, meas(i));
    end
end

File: docs/tutorials/live_scripts/02_computing_measures.py
# %% [markdown]
# # Tutorial 2 — Computing All 26 Measures
#
# A detailed walkthrough of each block of the 26-measure array, plus how
# to call individual measures directly.

# %% [markdown]
# ## Setup

# %%
import numpy as np
from metasignal import stdpy

MEASURE_NAMES = [
    "meta-d'", "AUC2", "Gamma", "Phi", "DeltaConf",
    "M-Ratio", "AUC2-Ratio", "Gamma-Ratio", "Phi-Ratio", "DeltaConf-Ratio",
    "M-Diff", "AUC2-Diff", "Gamma-Diff", "Phi-Diff", "DeltaConf-Diff",
    "metaNoise", "metaUncertainty", "d'", "c", "mean_conf",
    "logL", "AIC", "BIC", "AICc", "k", "n",
]
N_MEAS = 26

rng = np.random.default_rng(0)
n_trials, n_ratings = 400, 4

stim = rng.choice([0, 1], n_trials)
resp = np.where(rng.random(n_trials) < 0.78, stim, 1 - stim)
correct = stim == resp
conf = np.where(
    correct,
    rng.integers(3, n_ratings + 1, n_trials),
    rng.integers(1, 3, n_trials),
)
print("Data ready:", n_trials, "trials,", n_ratings, "ratings")

# %% [markdown]
# ## Block 1 — Metacognitive sensitivity (indices 0–4)
#
# These five measures ask: *how well does confidence track accuracy?*

# %%
meas = stdpy.compute_all_measures(stim, resp, conf, n_ratings=n_ratings)

labels = ["meta-d'", "AUC2", "Gamma", "Phi", "DeltaConf"]
for i, name in enumerate(labels):
    print(f"  {name:<12} = {meas[i]:.4f}")

# %% [markdown]
# ## Block 2 & 3 — Efficiency ratios and differences (indices 5–14)
#
# Normalise observed metacognition by the *expected* performance of an ideal
# observer with the same d'. Removes spurious dependence on task difficulty.

# %%
nr_s1, nr_s2 = stdpy.trials_to_counts(stim, resp, conf, n_ratings=n_ratings)

expected   = stdpy.sdt_expect_conf(nr_s1, nr_s2)
nr_s1_exp  = np.array(expected["nR_S1_exp"])
nr_s2_exp  = np.array(expected["nR_S2_exp"])

auc2_obs = stdpy.compute_type2_auc(nr_s1, nr_s2)
auc2_exp = stdpy.compute_type2_auc(nr_s1_exp, nr_s2_exp)

print(f"AUC2 observed = {auc2_obs:.4f}")
print(f"AUC2 ideal    = {auc2_exp:.4f}")
print(f"AUC2-Ratio    = {meas[6]:.4f}  (obs / ideal)")
print(f"AUC2-Diff     = {meas[11]:.4f} (obs − ideal)")

# %% [markdown]
# ## Individual measure functions

# %%
gamma = stdpy.compute_gamma(nr_s1, nr_s2)
phi   = stdpy.compute_phi(nr_s1, nr_s2)
dc    = stdpy.compute_delta_conf(nr_s1, nr_s2)

print(f"Gamma      = {gamma:.4f}")
print(f"Phi        = {phi:.4f}")
print(f"DeltaConf  = {dc['delta_conf']:.4f}")

result = stdpy.fit_meta_d_mle(nr_s1, nr_s2)
print(f"meta_da    = {result['meta_da']:.4f}")
print(f"M_ratio    = {result['M_ratio']:.4f}")

# %% [markdown]
# ## Block 4 — Meta-noise and meta-uncertainty (indices 15–16)

# %%
noise_res = stdpy.compute_meta_noise(stim, resp, conf, n_ratings=n_ratings)
uncert    = stdpy.compute_meta_uncertainty(stim, resp, conf, n_ratings=n_ratings)

print(f"metaNoise       = {noise_res['meta_noise']:.4f}")
print(f"metaUncertainty = {uncert:.4f}")

# %% [markdown]
# ## Full 26-measure summary

# %%
print(f"{'Index':<6} {'Measure':<20} {'Value':>10}")
print("-" * 40)
for i, (name, val) in enumerate(zip(MEASURE_NAMES, meas)):
    vstr = f"{val:10.4f}" if not np.isnan(val) else "       NaN"
    print(f"[{i:2d}]   {name:<20} {vstr}")
